### PR TITLE
Use CTE instead of inheritance views when generating ranges for types

### PR DESCRIPTION
**Pull request by dnwpark**
If a type or pointer has children and the compiler is not using EXPLAIN, use a CTE instead of the current inheritance views.

Also remove the `expand_inhview` debug flag.

### CODE REVIEW

**Comment by dnwpark** [comment]
The flag is a bit awkward, the end goal is to have two flags, `is_explain` and `use_type_inheritance_ctes` while removing the current `expand_inhviews`.

**Comment by aljazerzen** [comment]
@dnwpark, my PRs are merged, SQL now uses inheritance CTEs

**Comment by dnwpark** [comment]
For example, given:
```
  type Foo {
    n -> int64;
  }
  type Foo2 extending Foo;
  type Bar {
    foo -> Foo {
      n -> str
    }
  }
  type Bar2 extending Bar;
```

The query `select Foo { n }` produces the sql:
```sql
WITH 
"t_default::Foo~1" AS NOT MATERIALIZED 
    ((
    (
        FROM edgedbpub."default::Foo" AS "Foo~2"
        SELECT/*<pg.SelectStmt at 0x7fe6befd7aa0>*/
            "Foo~2".id AS "Foo_value~1",
            "Foo~2"."n" AS "n_serialized~1"
    ) union ALL (
        FROM edgedbpub."default::Foo2" AS "Foo2~1"
        SELECT/*<pg.SelectStmt at 0x7fe6be4e7ad0>*/
            "Foo2~1".id AS "Foo_value~2",
            "Foo2~1"."n" AS "n_serialized~2"
    )
    ))
FROM (
    FROM "t_default::Foo~1" AS "t_default::Foo~2"
    SELECT/*<pg.SelectStmt at 0x7fe6be4e7260>*/
        "t_default::Foo~2"."Foo_value~1" AS "Foo_value~3",
        "t_default::Foo~2"."n_serialized~1" AS "n_serialized~3"
) AS "q~1"
SELECT/*<pg.SelectStmt at 0x7fe6be0354f0>*/
    ROW((
        SELECT/*<pg.SelectStmt at 0x7fe6be4e6420>*/
            "q~1"."n_serialized~3" AS "n_serialized~4"
    )) AS "Foo_serialized~1"
```

The query `select Foo2 { n }` produces the sql:
```sql
FROM edgedbpub."default::Foo2" AS "Foo2~2"
SELECT/*<pg.SelectStmt at 0x7fe6bd8bee70>*/
    ROW((
        SELECT/*<pg.SelectStmt at 0x7fe6be4e66c0>*/
            "Foo2~2"."n" AS "n_serialized~1"
    )) AS "Foo2_serialized~1"
```

The query `select Bar { foo : { n, @n } }` produces the sql:
```sql
WITH 
"t_default::Bar~1" AS NOT MATERIALIZED 
    ((
    (
        FROM edgedbpub."default::Bar" AS "Bar~2"
        SELECT/*<pg.SelectStmt at 0x7fe6bd8be630>*/
            "Bar~2".id AS "Bar_value~1",
            "Bar~2"."foo" AS "foo_identity~1"
    ) union ALL (
        FROM edgedbpub."default::Bar2" AS "Bar2~1"
        SELECT/*<pg.SelectStmt at 0x7fe6be4e61b0>*/
            "Bar2~1".id AS "Bar_value~2",
            "Bar2~1"."foo" AS "foo_identity~2"
    )
    )),
"t_default::Foo~1" AS NOT MATERIALIZED 
    ((
    (
        FROM edgedbpub."default::Foo" AS "Foo~1"
        SELECT/*<pg.SelectStmt at 0x7fe6be201fa0>*/
            "Foo~1".id AS "Foo_identity~1",
            "Foo~1"."n" AS "n_serialized~1"
    ) union ALL (
        FROM edgedbpub."default::Foo2" AS "Foo2~1"
        SELECT/*<pg.SelectStmt at 0x7fe6be201550>*/
            "Foo2~1".id AS "Foo_identity~2",
            "Foo2~1"."n" AS "n_serialized~2"
    )
    )),
"t_default::__|foo@default|Bar~1" AS NOT MATERIALIZED 
    ((
    (
        FROM edgedbpub."foo" AS "foo~1"
        SELECT/*<pg.SelectStmt at 0x7fe6be202600>*/
            "foo~1".source AS source,
            "foo~1".target AS target,
            "foo~1"."@n" AS "n_serialized~5"
    ) union ALL (
        FROM edgedbpub."foo" AS "foo~2"
        SELECT/*<pg.SelectStmt at 0x7fe6be200650>*/
            "foo~2".source AS source,
            "foo~2".target AS target,
            "foo~2"."@n" AS "n_serialized~6"
    )
    ))
FROM (
    FROM "t_default::Bar~1" AS "t_default::Bar~2"
    SELECT/*<pg.SelectStmt at 0x7fe6be4e5d30>*/
        "t_default::Bar~2"."Bar_value~1" AS "Bar_value~3",
        "t_default::Bar~2"."foo_identity~1" AS "foo_identity~3"
) AS "q~1"
SELECT/*<pg.SelectStmt at 0x7fe6bd72c890>*/
    ROW((
        FROM LATERAL (
            FROM LATERAL (
                SELECT/*<pg.SelectStmt at 0x7fe6beb596d0>*/
                    "q~1"."foo_identity~3" AS "foo_identity~4"
            ) AS "q~2"
            JOIN LATERAL (
                FROM "t_default::Foo~1" AS "t_default::Foo~2"
                SELECT/*<pg.SelectStmt at 0x7fe6be201d90>*/
                    "t_default::Foo~2"."Foo_identity~1" AS "Foo_identity~3",
                    "t_default::Foo~2"."n_serialized~1" AS "n_serialized~3"
            ) AS "q~3"
                ON ("q~2"."foo_identity~4" = "q~3"."Foo_identity~3")
            SELECT/*<pg.SelectStmt at 0x7fe6be4e4830>*/
                "q~3"."Foo_identity~3" AS "foo_value~1",
                ((
                    SELECT/*<pg.SelectStmt at 0x7fe6be2023f0>*/
                        "q~3"."n_serialized~3" AS "n_serialized~4"
                ), (
                    FROM (
                        FROM "t_default::__|foo@default|Bar~1" AS "t_default::__|foo@default|Bar~2"
                        SELECT/*<pg.SelectStmt at 0x7fe6be2037a0>*/
                            "t_default::__|foo@default|Bar~2".source AS source,
                            "t_default::__|foo@default|Bar~2".target AS target,
                            "t_default::__|foo@default|Bar~2"."n_serialized~5" AS "n_serialized~7"
                    ) AS "q~4"
                    SELECT/*<pg.SelectStmt at 0x7fe6be201070>*/
                        "q~4"."n_serialized~7" AS "n_serialized~8"
                    WHERE
                        (("q~1"."Bar_value~3" = "q~4".source) AND 
                         ("q~2"."foo_identity~4" = "q~4".target))
                )) AS "foo_serialized~1"
        ) AS "Foo_foo~3"
        SELECT/*<pg.SelectStmt at 0x7fe6be4e75c0>*/
            "Foo_foo~3"."foo_serialized~1" AS "foo_serialized~2"
    )) AS "Bar_serialized~1"
```

The query `select Bar2 { foo : { n, @n } }` produces the sql:
```sql
WITH 
"t_default::Foo~1" AS NOT MATERIALIZED 
    ((
    (
        FROM edgedbpub."default::Foo" AS "Foo~1"
        SELECT/*<pg.SelectStmt at 0x7fe6bd6a8110>*/
            "Foo~1".id AS "Foo_identity~1",
            "Foo~1"."n" AS "n_serialized~1"
    ) union ALL (
        FROM edgedbpub."default::Foo2" AS "Foo2~1"
        SELECT/*<pg.SelectStmt at 0x7fe6bd6a9fd0>*/
            "Foo2~1".id AS "Foo_identity~2",
            "Foo2~1"."n" AS "n_serialized~2"
    )
    ))
FROM edgedbpub."default::Bar2" AS "Bar2~2"
SELECT/*<pg.SelectStmt at 0x7fe6be4e45f0>*/
    ROW((
        FROM LATERAL (
            FROM LATERAL (
                SELECT/*<pg.SelectStmt at 0x7fe6becb5a60>*/
                    "Bar2~2"."foo" AS "foo_identity~1"
            ) AS "q~1"
            JOIN LATERAL (
                FROM "t_default::Foo~1" AS "t_default::Foo~2"
                SELECT/*<pg.SelectStmt at 0x7fe6bd6a8920>*/
                    "t_default::Foo~2"."Foo_identity~1" AS "Foo_identity~3",
                    "t_default::Foo~2"."n_serialized~1" AS "n_serialized~3"
            ) AS "q~2"
                ON ("q~1"."foo_identity~1" = "q~2"."Foo_identity~3")
            SELECT/*<pg.SelectStmt at 0x7fe6bec1b9b0>*/
                "q~2"."Foo_identity~3" AS "foo_value~1",
                ((
                    SELECT/*<pg.SelectStmt at 0x7fe6bd6a8b00>*/
                        "q~2"."n_serialized~3" AS "n_serialized~4"
                ), (
                    FROM edgedbpub."foo" AS "foo~1"
                    SELECT/*<pg.SelectStmt at 0x7fe6bd6aaa80>*/
                        "foo~1"."@n" AS "n_serialized~5"
                    WHERE
                        (("Bar2~2".id = "foo~1".source) AND 
                         ("q~1"."foo_identity~1" = "foo~1".target))
                )) AS "foo_serialized~1"
        ) AS "Foo_foo~3"
        SELECT/*<pg.SelectStmt at 0x7fe6bec190a0>*/
            "Foo_foo~3"."foo_serialized~1" AS "foo_serialized~2"
    )) AS "Bar2_serialized~1"
```

File: edb/pgsql/compiler/relctx.py [hunk 1]
```diff
@@ -1640,49 +1641,104 @@ def range_for_material_objtype(
...
+            # When we are compiling a query for EXPLAIN, expand out type
+            # references to an explicit union of all the types, rather than
+            # using a CTE. This allows postgres to actually give us back the
+            # alias names that we use for relations, which we use to track which
+            # parts of the query are being referred to.
+            not ctx.env.use_inheritance_ctes
+
+            # Don't use CTEs if there is no inheritance. (ie. There is only a
+            # single material type)
+            or len(typeref.children) == 0
```
Comment (msullivan) [inline edb/pgsql/compiler/relctx.py:1662]: I think the `== 0` case should maybe be folded into the above condition, so we don't add an extra select around it for no reason

File: edb/pgsql/compiler/relctx.py [hunk 2]
```diff
@@ -2039,193 +2076,420 @@ def range_for_ptrref(
...
-    lrefs: dict[irast.PointerRef, list[irast.PointerRef]]
-    if (
-        ctx.env.expand_inhviews
-        and include_descendants
-        and not for_mutation
+    include_descendants = (
+        not ptrref.union_is_exhaustive
 
         # HACK: This is a workaround for #4491
         and ptrref.out_source.name_hint.module not in {'sys', 'cfg'}
```
Comment (msullivan) [inline edb/pgsql/compiler/relctx.py:2100]: We should drop these checks

File: edb/pgsql/compiler/relctx.py [hunk 3]
```diff
@@ -2039,193 +2076,420 @@ def range_for_ptrref(
...
+        # single ptrref)
+        or len(ptrref_descendants) <= 1
+    ):
+        descendant_selects = _selects_for_ptrref_descendants(
+            ptrref_descendants,
+            output_cols=output_cols,
+            path_id=path_id,
+            ctx=ctx,
+        )
+        descentant_ops = [
```
Comment (msullivan) [inline edb/pgsql/compiler/relctx.py:2183]: spelling

**Review by msullivan — APPROVED** [review]
This looks great!

File: edb/edgeql/compiler/options.py [hunk 4]
```diff
@@ -77,10 +77,15 @@ class GlobalCompilerOptions:
...
+    #: Should the backend compiler expand inheritance CTEs in place.
+    #: This is needed by EXPLAIN to maintain alias names in
     #: the query plan.
-    expand_inhviews: bool = False
+    is_explain: bool = False
+
+    #: Should type inheritance be expanded using CTEs.
+    #: When not explaining CTEs can be used to provide access to a type and its
+    #: descendents.
+    use_inheritance_ctes: bool = True
```
[the remark this answers is not in the document]
Comment (dnwpark) [reply edb/edgeql/compiler/options.py:88]: In `server/compiler/compiler.py`, this flag is set to false for explain, bootstrapping and schema reflection.

[the remark this answers is not in the document]
Comment (msullivan) [reply edb/edgeql/compiler/options.py:88]: Does it need to be set to false for bootstrapping and schema reflection? We need to stop using inheritance views for those too

File: edb/pgsql/compiler/relctx.py [hunk 5]
```diff
@@ -1740,34 +1739,6 @@ def range_for_material_objtype(
                 rvar = rvar_for_rel(
                     sctx.rel, lateral=lateral, typeref=typeref, ctx=sctx)
 
-    else:
-        assert not typeref.is_view, "attempting to generate range from view"
```
Comment (msullivan) [inline edb/pgsql/compiler/relctx.py:1688]: Could we preserve this assert in the new path?

File: edb/pgsql/compiler/context.py [hunk 6]
```diff
@@ -246,12 +246,22 @@ class CompilerContextLevel(compiler.ContextLevel):
...
 
     #: A set of type CTEs currently being generated
     pending_type_ctes: Set[RewriteKey]
 
+    #: CTEs representing types and their inherited types
+    type_inheritance_ctes: Dict[uuid.UUID, pgast.CommonTableExpr]
+
+    # Type and type inheriance CTEs in creation order. This ensures type CTEs
+    # referring to other CTEs are in the correct order.
+    ordered_inheritance_ctes: list[pgast.CommonTableExpr]
```
[the remark this answers is not in the document]
Comment (dnwpark) [reply edb/pgsql/compiler/context.py:263]: Unfortunately, the two kinds of type ctes (rewrites and inheritance) can depend on each other and may have an interleaved order. The pointers ctes are entirely separate though, and I've separated them now.

File: edb/pgsql/compiler/relctx.py (hunk 5 above)
Comment (dnwpark) [reply edb/pgsql/compiler/relctx.py:1688]: I put it in `_table_from_typeref` but I think it's reasonable to have it more ahead in the process.

File: edb/pgsql/compiler/relctx.py [hunk 7]
```diff
@@ -1779,6 +1807,103 @@ def range_for_material_objtype(
...
+    assert isinstance(typeref.name_hint, sn.QualName)
+
+    aspect = 'table'
+
+    if (
+        include_descendants
+        and not for_mutation
+        and typeref.name_hint.module in {'cfg', 'sys', 'schema'}
+    ):
+        aspect = 'inhview'
```
Comment (dnwpark) [inline edb/pgsql/compiler/relctx.py:1881]: @msullivan Thoughts on this? I would like to add a comment here, but I'm not entirely sure how best to explain this branch beyond something vague like, "these modules use views to implement inheritance".

File: edb/edgeql/compiler/typegen.py [hunk 8]
```diff
@@ -401,13 +401,7 @@ def type_to_typeref(
...
         expr_type is s_types.ExprType.Update
         or expr_type is s_types.ExprType.Delete
-        or (
-            (
-                env.options.is_explain or
-                env.options.use_inheritance_ctes
-            )
-            and isinstance(t, s_objtypes.ObjectType)
-        )
+        or isinstance(t, s_objtypes.ObjectType)
```
Comment (msullivan) [inline edb/edgeql/compiler/typegen.py:404]: We can collapse out the `expr_type` conditions--those can only hold if the type is an object anyway.

**Review by msullivan — APPROVED** [review]
Great work on this!

**Pull request merged**